paddle/fluid/platform/errors.h

```diff
@@ -0,0 +1,104 @@
+/* Copyright (c) 2019 PaddlePaddle Authors. All Rights Reserved.
+
+licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License. */
+
+#pragma once
+
+#include <memory>
+#include <stdexcept>
+#include <string>
+#include <tuple>
+#include <type_traits>
+
+#include "paddle/fluid/platform/error_codes.pb.h"
+#include "paddle/fluid/string/printf.h"
+
+namespace paddle {
+namespace platform {
+
+typedef ::paddle::platform::error::Code Code;
+
+class ErrorSummary {
+ public:
+  // Note(chenweihang): Final deprecated constructor
+  //   This constructor is only used to be compatible with
+  //   current existing no error message PADDLE_ENFORCE_*
+  ErrorSummary() {
+    state_ = std::unique_ptr<State>(new State);
+    state_->code = paddle::platform::error::LEGACY;
+    state_->msg =
+        "Paddle internal Check failed. (Please help us create a new issue, "
+        "here we need to find the developer to add a user friendly error "
+        "message)";
+  }
+
+  // Note(chenweihang): Final deprecated constructor
+  //   This constructor is used to be compatible with
+  //   current existing untyped PADDLE_ENFORCE_*
+  //   PADDLE_ENFORCE
+  template <typename... Args>
+  explicit ErrorSummary(Args... args) {
+    state_ = std::unique_ptr<State>(new State);
+    state_->code = paddle::platform::error::LEGACY;
+    state_->msg = paddle::string::Sprintf(args...);
+  }
+
+  // Note(chenweihang): Recommended constructor
+  explicit ErrorSummary(Code code, std::string msg) {
+    state_ = std::unique_ptr<State>(new State);
+    state_->code = code;
+    state_->msg = msg;
+  }
+
+  Code code() const { return state_->code; }
+
+  const std::string& error_message() const { return state_->msg; }
+
+  std::string ToString() const;
+
+ private:
+  struct State {
+    Code code;
+    std::string msg;
+  };
+
+  std::unique_ptr<State> state_;
```

Comment:
Why `unique_ptr`? Is it ok to just use `State` instead of `std::unique_ptr<State>` ?

Reply:
done